pkg/status/status.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package status
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/status/internal/pb"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+	"github.com/ethersphere/bee/pkg/topology/depthmonitor"
+)
+
+// loggerName is the tree path name of the logger for this package.
+const loggerName = "status"
+
+const (
+	protocolName    = "status"
+	protocolVersion = "1.0.0"
+	streamName      = "status"
+)
+
+// Snapshot is the current snapshot of the system.
+type Snapshot struct {
+	Peer          swarm.Address
+	Proximity     uint8
+	ReserveSize   uint64
+	PullsyncRate  float64
+	StorageRadius uint8
+}
+
+// Service is the status service.
+type Service struct {
+	logger       log.Logger
+	streamer     p2p.Streamer
+	topologyIter topology.PeerIterator
+
+	reserve depthmonitor.ReserveReporter
+	sync    depthmonitor.SyncReporter
+	radius  postage.RadiusReporter
+}
+
+// Snapshots returns the current status snapshot of this node or connected peers.
+func (s *Service) Snapshots(ctx context.Context, connectedPeers bool) ([]*Snapshot, error) {
+	snapshots := []*Snapshot{{
+		ReserveSize:   s.reserve.ReserveSize(),
+		PullsyncRate:  s.sync.SyncRate(),
+		StorageRadius: s.radius.StorageRadius(),
+	}}
+
+	if !connectedPeers {
+		return snapshots, nil
+	}
+
+	peerFunc := func(address swarm.Address, po uint8) (stop, jumpToNext bool, err error) {
+		select {
+		case <-ctx.Done():
+			return true, false, nil
+		default:
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+
+		ss, err := s.requestStatusSnapshot(ctx, address)
+		if err != nil {
+			s.logger.Debug("cannot get status snapshot for peer", "peer_address", address, "error", err)
+			return false, true, nil
```

Comment:
I am not sure if this is correct usage of `jumpToNext`. This is used to jump to next bin I think. Can you check? I think this should be false here.

Reply:
this would skip to the next bin, so we cannot return true here. if we cannot get the status, we can have a new error field in the response that would be filled on error

Reply:
Done.